CSV Changes with Redux Toolkit and Formatting Updates

mikedivine opened the pull request:
# Description

Removed CSV container and split into separate CSV pages.  Uses Create Meter Modal in Readings and seems to be working with Redux.

@rfrost30 has helped contribute on this PR.

Fixes #1198

## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement]

## Limitations

None

mikedivine committed:
fixed CSV with previous changes

mikedivine committed:
Finished formatting formatting CSV and updating tooltip data, tested with csv uploads manually and looks to be working

mikedivine committed:
fixed csvUploadDefaults to remove create meter option

mikedivine committed:
Merge branch 'development' into 1198-CSV

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -26,315 +33,477 @@ function range(lower: number, upper: number): number[] {
+										</Label>
+									</Col>
+								</Row>
+								<Row>
```
I think we should change the refresh behavior from the past (you correctly mirrored the old code). I think this should only have a "Refresh Readings" box so the two become one. The rationale is the graphing code needs both now and the refresh is moderately fast so doing both should not be too bad. The function in src/server/services/refreshAllReadingViews.js could be used to do this rather than two calls.

mikedivine replied:
Done. I have now changed it to only use one parameter of Refresh Readings and the Server API now uses the function in src/server/services/refreshAllReadingViews.js to do this.

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -26,315 +33,477 @@ function range(lower: number, upper: number): number[] {
+		e.preventDefault();
+		if (selectedFile) {
+			try {
+				const msg = await submitReadings(selectedFile);
```
This is continuing to use the old style routes outside the RTK (new in src/client/app/redux/api). I think OED should probably go to the new system to be consistent and to allow other actions such as invalidating the current readings so you don't have to refresh the web browser after an upload. This could be done now or an issue opened if people agree.

mikedivine replied:
Rewrote this api to have readings invalidated once new readings are added.  An issue can still be added so in the future it is written to add the data without invalidating.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
-		);
-	}
+	return (
+		<Container>
```
This begins shifted down the page so it is not near the title. Is it easy to get to start at the top of the page?

mikedivine replied:
Looks like this is resolved now.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
+		<Container>
+			<TooltipHelpComponent page='help.csv.header' />
+			<Form onSubmit={handleSubmit}>
+				<Row className="justify-content-md-center">
```
I see the parallel with CSV Readings with Row/Col. Could you explain if these are also helping with the page format? I just want to understand.

mikedivine replied:
Yes, it just helps with alignments and spacing, etc.

huss replied:
I'm still unclear so could someone help me understand?

mikedivine replied:
These are ReactStrap elements I saw used on the other components and they just make sure everything is aligned in rows and columns. This could be done with div elements if you prefer or another method if you have a suggestion, just let me know if you have a preference.

mikedivine replied:
OK, I got rid of the Row/Col except for outside to center everything inside a column and then rows and columns for the checkboxes.

huss commented on src/client/app/components/TooltipHelpComponent.tsx:
```diff
@@ -47,6 +47,8 @@ export default function TooltipHelpComponent(props: TooltipHelpProps) {
 		'help.admin.unitview': { link: `${helpUrl}/adminUnitViewing.html` },
 		'help.admin.user': { link: `${helpUrl}/adminUser.html` },
 		'help.csv.header': { link: `${helpUrl}/adminDataAcquisition.html` },
```
I see this tag is used for the page label but I don't see it being used as a tooltip/translation. Also, data.ts has it in two languages but not French. Should this tag be removed? If so, the documentation needs to be checked that it gives what is needed on the linked pages (someone else can do the docs).

mikedivine replied:
Yes, I believe this can be removed as a help link but the documentation can stay because it has useful information outside of using the site to upload data.

huss replied:
I see it gone from data.ts but it is still in this file. Is it needed?

mikedivine replied:
Oh, I see, removed.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
-		this.handleSetMeterName = this.handleSetMeterName.bind(this);
-		this.fileInput = React.createRef<HTMLInputElement>();
-	}
+interface MetersCSVUploadComponentProps {}
```
This probably is needed but I wonder if there is another way than an empty interface.

mikedivine replied:
no longer needed, was when we had to pass props from a container

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
+interface MetersCSVUploadComponentProps {}
+
+const MetersCSVUploadComponent: React.FC<MetersCSVUploadComponentProps> = () => {
+	const dispatch = useDispatch();
```
Most of the code base now uses useAppDispatch and wondewr if this can too. This may relate to updating for RTK.

mikedivine replied:
Fixed to use AppDispatch

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
+
+const MetersCSVUploadComponent: React.FC<MetersCSVUploadComponentProps> = () => {
+	const dispatch = useDispatch();
+	const[ meterData, setMeterData] = React.useState({
```
I'm not taking a position on this but many places use a defaultState/defaultValues (name not consistent) to define and initialize state. The CSV Readings does it too. Should OED prefer one? Should the code try to be consistent?

mikedivine replied:
Yes, default state and values now used.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
+	};
 
-	}
+	const handleFileChange = (file : File | null) => {
```
OED tries to use the default VSC formatting (command: format document). I think this should be applied to all files to conform (here no space before the :).

mikedivine replied:
done

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
+		headerRow: false,
+		update: false
+	});
+	const [ file, setFile ] = React.useState<File | null>(null);
```
I'm curious about using state vs a local variable when no React/Redux is used (such as useEffect). I don't know which OED has used but I think it is variables. Something to discuss/decide.

mikedivine replied:
hmmm, not sure I understand, let's discuss in meeting.

huss replied:
Is this resolved?

mikedivine replied:
Is this referring to the file reference or this: const [selectedFile, setSelectedFile] = React.useState<File | null>(null);

mikedivine replied:
It's working this way.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
+	const [ file, setFile ] = React.useState<File | null>(null);
+	const fileInput = React.useRef<HTMLInputElement>(null);
+	const resetApiCache = () => {
+		dispatch(baseApi.util.invalidateTags(['MeterData']));
```
I suspect this would not be needed here if the RTK was used for routing.

mikedivine replied:
Yes, correct, I believe we will have to create a seperate issue for this since all the CSV processing is done on the server side currently, so I don't see an easy way to use the RTK

mikedivine replied:
I rewrote the api to update redux with the meters adding, although you may need to look and see if my method is ideal for you or what your thoughts would be.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
-				showSuccessNotification('<h1>SUCCESS</h1>The meter upload was a success.');
+	const handleSubmit = async(e: React.MouseEvent<HTMLFormElement>) => {
+		e.preventDefault();
+		if(file) {
```
Some places in the newer OED forms it can check if the required elements are done and not enable the submit until that is done. Is that possible with a file picker?

mikedivine replied:
now implemented

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -26,315 +33,477 @@ function range(lower: number, upper: number): number[] {
 	return arr;
 }
 
```
It seemed the message with info on upload status was lost due to a web page refresh. Need to look into.

mikedivine replied:
this has been resolved.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
 		}
-		// Refetch meters details by invalidating its api cache.
-		this.props.resetApiCache();
+		resetApiCache();
```
Could there be a comment if this is needed. I would also like to know why it is here.

mikedivine replied:
added a comment in the code:
	// This is needed until a React CSV API is created or integrated with the Redux meter API

huss replied:
Maybe add a comment, even if similar to the one above about the React CSV API.

mikedivine replied:
Will do.

mikedivine replied:
I have updated the frontend CSV api to now send new meters using dispatch to update the store. If it's to update meters, it invalidates all meters as discussed with you previous. This is how the current redux api does it, but we weren't sure why exactly. If it changes in the future, we can adjust the api accordingly.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -580,6 +584,7 @@ const LocaleTranslationData = {
 		"csv": "CSV",
 		"csv.file": "Fichier CSV",
+		"csv.clear.button": "Forme claire",
 		"csv.common.param.gzip": "GzipComment",
```
I'm just noticing now but is the French string for gzip correct?

mikedivine replied:
Should just be Gzip like the others, fixed

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
+				<Row className='justify-content-md-center'>
+					<Col md='auto'>
+						<FormFileUploaderComponent
+							formText='csv.upload.meters'
+							onFileChange={handleFileChange}
```
Do people think there should be a second string saying meter upload when in the title above?

mikedivine replied:
Yeah, that is a good point.  I would be inclined to say it is unnecessary with the title above.

mikedivine replied:
I removed it.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
+									<Col>
+										<Label for='gzip'>
+											<div style={checkBox}>
+												<div>
```
Is this div needed? If it is for spacing maybe use the checkBox style to add that.

mikedivine replied:
you are correct, not needed, will remove.

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -3,14 +3,21 @@
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 import * as React from 'react';
```
In all the files it would be nice to use the VSC command "organize imports" that will sort them in a standard way.

mikedivine replied:
done, thanks for the tip

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -26,315 +33,477 @@ function range(lower: number, upper: number): number[] {
+									{translate('meter.readingDuplication')}
+								</div>
+								<Input type='select' id='duplications' name='duplications' value={selectedMeter?.readingDuplication || ''} onChange={handleChange}>
+									{range(1, 10).map(i => (
```
I like the way this is done. The recent change to the meter create/edit pages lists each one. Are you willing to roll this to those pages? Maybe put the range functions into a general util directory.

mikedivine replied:
This range function was written in the original combined CSV components, says by lindavin. But yes, I can add this in utilities and use this function in the create and edit pages.

mikedivine replied:
I have now added this to utilities and is now used in create and edit Meters components.

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -26,315 +33,477 @@ function range(lower: number, upper: number): number[] {
-		this.props.setMeterName(MODE.readings, target.value);
-	}
+/**
+ * Defines the CSV Readings page card view
```
Is this a card view? Maybe put this type of comment in the CSV meters file.

mikedivine replied:
Not a card view, oopsie, removed that portion of the comment

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -26,315 +33,477 @@ function range(lower: number, upper: number): number[] {
-										<option value={BooleanMeterTypes.false}> {translate('BooleanMeterTypes.false')} </option>
-									</Input>
+						<br />
+						<CreateMeterModalComponent onCreateMeter={handleCreateMeter}/>
```
I tried being logged in with the CSV role and was able to create a new meter. However, only an admin should be able to do that and if the CSV user goes to the meter page they cannot. I'm not sure what is up but it is an issue.

Also, if a CSV user cannot create a meter then it might be good to revisit the logic above about a newly created meter that is not visible to the CSV user.

mikedivine replied:
If a CSV user can create meters by uploading csv data then I find it odd that they cannot manually do so. Seems like a contradiction, but I definitely can implement where the create meter button is disabled if they are not an admin if that is preferred.

mikedivine replied:
This has now been set to not allow CSV users to create a meter manually and still only see visible meters to upload readings for.

mikedivine replied:
Also they can no longer access the CSV Meters page to upload meters

huss commented on src/client/app/components/HeaderButtonsComponent.tsx:
```diff
@@ -174,8 +174,15 @@ export default function HeaderButtonsComponent() {
 								style={state.csvViewableLinkStyle}
 								disabled={state.shouldCSVButtonDisabled}
```
This is not working as I would expect. It is supposed to gray out the current page. Given this, I think there needs to be state for each CSV page and not just one. NOte the route seems wrong above and that should be part of this change.

mikedivine replied:
I did not see this first go around.  This is now updated and seems to work properly.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -86,35 +86,38 @@ const LocaleTranslationData = {
 		"create.unit": "Create a Unit",
 		"csv": "CSV",
-		"csv.file": "CSV File",
+		"csvMeters": "CSV Meters",
```
Many of the changes in English are not reflected in the other languages.

mikedivine replied:
Ok, I believe these should all be addressed now.

huss reviewed (commented):
Thanks to @mikedivine and anyone else involved in this PR. Overall, it works as desired. I am releasing this as comments to get my thoughts to the team. I suspect we will want to discuss at an upcoming meeting. I still need to do some more careful reading/testing but maybe that can wait until these comments are decided. As always, I appreciate your thoughts and I'm here to clarify, esp. given my thoughts were released quickly.

Commit:
fixed MetersCSVUpload to align to top of the page

mikedivine committed:
formatted imports and file, fixed tooltip headings, fixed menu items grayed out if on that page

mikedivine committed:
only show create meter button if admin, made range a utlity and updated create and edit meter with it. Fixed bug where the form metername wasn't being set on a newly created meter

mikedivine committed:
updated translation data

mikedivine committed:
updated CSV pipeline to use meterIdentifier instead of meterName to upload readings and create meters via csv

Commit:
fixed MetersCSVUpload to align to top of the page

Commit:
updated translation to include meter identifier

mikedivine committed:
Updated CSV Readings to use only Refresh Readings and not Refresh hourly and daily.  Updated the API to only use that option and refreshAllReadingViews if requested.

mikedivine committed:
Updated CSV Readings and Meters to include validation and made fileReference optional in the file upload component.

mikedivine committed:
fixes to allow Gzip files as well as csv files

mikedivine committed:
added some comments to readings component

huss commented on src/client/app/translations/data.ts:
```diff
@@ -230,7 +235,8 @@ const LocaleTranslationData = {
 		"help.admin.unitview": "This page shows information on units. Please visit {link} for further details and information.",
 		"help.admin.user": "This page allows admins to view and edit users. Please visit {link} for further details and information.",
-		"help.csv.header": "This page allows certain users to upload meters and readings via a CSV file. Please visit {link} for further details and information.",
+		"help.csv.meters": "This page allows certain users to upload meters via a CSV file. Please visit {link} for further details and information.",
```
I think this should be "allows admins to" since we decided to limit to admins.

mikedivine replied:
Got it, fixed

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -3,338 +3,490 @@
+												type='select'
+												id='cumulativeReset'
+												name='cumulativeReset'
+												value={getBooleanMeterType(selectedMeter?.cumulative)}
```
I think this should be cumulativeReset.

mikedivine replied:
yes, fixed

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -3,338 +3,490 @@
+									type='select'
+									id='timeSort'
+									name='timeSort'
+									value={selectedMeter?.timeSort || ''}
```
I think all the input with state similar to this have an issue. The handle updates the readingsData state but the value shown is the meter value. I think this relates to another concern I had that when you change a meter you only set the meterIdentifier in the readingsData state and not all the related values. You can fix as you want but here is a thought:
- Initially (as is done) the readingsData gets the default values. This means in most cases it will be the meter value. This, by definition, must change when the meter is selected (see next).
- When a meter is selected (which includes creation), the readingsData state is updated with the new meter's values. This does mean any selected items are reset.
- The value on item change is set to the readingsData value and not the meter. This avoids modifying the meter values but keeps the menus showing the current value.

mikedivine replied:
Understood, this makes a lot of sense and I will change it to use the method you described.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,192 @@
 
-}
-export default connect(null, mapDispatchToProps)(MetersCSVUploadComponent);
+	return (
```
I know we all discussed this but noting that only an admin will be allowed to create a meter.

mikedivine replied:
Yes, this is resolved now

huss commented on src/client/app/utils/range.ts:
```diff
@@ -0,0 +1,17 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Get array of numbers from a low and a high number
```
I liked the JSDoc in the original version in ReadingsCSVUploadComponent.tsx. It made it clear whether lower/higher was inclusive or exclusive. May not be needed given another comment.

mikedivine replied:
Got it, will do.

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -1,7 +1,6 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
-
 import * as _ from 'lodash';
```
I think most files have a blank line after the MPL notice so can it be added back.

mikedivine replied:
done

huss commented on src/client/app/utils/range.ts:
```diff
@@ -0,0 +1,17 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
I became curious if there was a standard way to do this. I found lots of solutions but I think these two seem the best choices:
1. lodash has a range function that does exactly what this function does.
2. [...Array(5).keys()]
Given we already use lodash, I think the first choice is the way to go. Just make sure to do the new import way to only import the needed function. This removes this file.

mikedivine replied:
got it, moving forward to use lodash

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,192 @@
 
```
This is a general comment so just sticking here. Given the time since development was merged with this work, I think it would be good to merge it again to be sure it is good with all the changes. (I just noted there are merge conflicts so this really needs to happen.)

huss commented on src/client/app/translations/data.ts:
```diff
@@ -579,35 +585,38 @@ const LocaleTranslationData = {
+		"csv.readings.param.length.variation": "Variation de longueur:",
+		"csv.readings.param.meter.name": "Nom du compteur:",
+		"csv.readings.param.meter.identifier": "Identifiant du compteur:",
+		"csv.readings.param.refresh.hourly.readings": "Actualiser les relevés horaires",
```
Is this translation used?

mikedivine replied:
no, removing

huss commented on src/server/services/pipeline-in-progress/validateReadings.js:
```diff
@@ -32,7 +32,7 @@ function validateReadings(arrayToValidate, conditionSet, meterName = undefined)
  * @param {Moment} maxDate inclusive latest acceptable date (won't be rejected)
  * @param {number} maxError maximum number of errors to be reported, ignore the rest
```
Not your doing but since you changed the name of the last parameter could you add it to the JDDoc? Same for ones below.

mikedivine replied:
How would I add something to the JSdoc?

huss replied:
You can put
`` * @param {string} meterName The name of the meter``
at the end of the param list. It will show up as part of the JSDoc in VSC (or if manually created which we don't do).

huss commented on src/server/services/csvPipeline/uploadReadings.js:
```diff
@@ -20,72 +17,21 @@ const Preferences = require('../../models/Preferences');
  */
 async function uploadReadings(req, res, filepath, conn) {
-	const { meterName, createMeter, headerRow, update, honorDst, relaxedParsing } = req.body; // extract query parameters
+	const { meterIdentifier, headerRow, update, honorDst, relaxedParsing } = req.body; // extract query parameters
```
I have a feeling that the changes in the PR to meterIdentifier impact what is sent by curl. Is that correct? If so, we need to be careful that the documentation is updated for what is actually sent.

mikedivine replied:
I don't believe it will change much for what is sent back because it just looks like:
  <h1>FAILURE</h1> <h2>It looks like the insert of the readings had issues with some or all of the readings where the processing of the readings returned these warning(s)/error(s):</h2>
  For reading #1 on meter pipe3 in pipeline: previous reading has value 0 start time 1970-01-01T00:00:00+00:00 end time 1970-01-01T00:00:00+00:00 and current reading has value 24 start time 2021-06-01T00:00:00+00:00 end time 2021-06-02T00:00:00+00:00 with timeSort increasing; duplications 1; cumulative true; cumulativeReset false; cumulativeResetStart 00:00:00; cumulativeResetEnd 23:59:59.999999; lengthGap 0; lengthVariation 0; onlyEndTime false
So "pipe3" could be referring to meterIdentifier or meterName, but the input of meterIdentifier is definitely a change so documentation needs to be updated with pertinent info.

huss commented on src/server/services/csvPipeline/uploadReadings.js:
```diff
@@ -20,72 +17,21 @@ const Preferences = require('../../models/Preferences');
-		log.warn('The create meter was set but the meter already existed for meter ' + meter.name);
-	}
-
+	console.log('The meter details from getByIdentifier: ' + meter);
```
Should this output be left?

mikedivine replied:
no, removed

huss commented on src/server/models/Meter.js:
```diff
@@ -166,6 +166,17 @@ class Meter {
 		return Meter.mapRow(row);
 	}
 
+	/**
```
Bad formatting.

mikedivine replied:
oops, fixed

huss commented on src/server/sql/meter/get_meter_by_identifier.sql:
```diff
@@ -0,0 +1,5 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This needs a DB migration. Let me know if you need help with this.

huss replied:
My mistake. Not needed.

huss commented on src/client/app/components/FormFileUploaderComponent.tsx:
```diff
@@ -3,14 +3,15 @@
-	reference: React.RefObject<HTMLInputElement>;
+	reference?: React.RefObject<HTMLInputElement>;
 	required: boolean;
 	formText: string;
```
This used to be used on this page but it seems removed. Should it stay? If not, all the uses of this interface should be updated.

mikedivine replied:
Are you talking about formText? Looks like this component is only used on uploadReadings and uploadMeters, I will remove it from the component

huss commented on src/client/app/translations/data.ts:
```diff
@@ -86,35 +86,40 @@ const LocaleTranslationData = {
-		"csv.readings.param.refresh.readings": "Refresh Daily Readings",
+		"csv.readings.param.length.gap": "Length Gap:",
+		"csv.readings.param.length.variation": "Length Variation:",
+		"csv.readings.param.meter.name": "Meter name:",
```
Is this key needed anymore?

mikedivine replied:
no, removing, fixed

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -3,338 +3,490 @@
-	}
+	const handleFileChange = (file: File | null) => {
+		setSelectedFile(file);
+		if (!file) return;
```
Please use { } on all if statements. Applies the CSV meter too. I might consider and if/else for all these cases since I think it is less prone to errors.

mikedivine replied:
Got it, fixed

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -3,338 +3,490 @@
+												id='cumulativeResetStart'
+												name='cumulativeResetStart'
+												onChange={handleChange}
+												required
```
I'm unsure what this required does since I don't see any warning if nothing is entered.

mikedivine replied:
yes, now removed

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -3,338 +3,490 @@
+						</FormGroup>
+						<div className='d-flex flex-row-reverse'>
+							<div className='p-3'>
+								<Button color='primary' type='submit' disabled={!isValidCSV}>
```
I like the disabled check. I also see the warning if you don't select a meter and submit. Would it make sense to not allow submit on no meter selected?

mikedivine replied:
Got it, the required on the file input doesn't allow to submit, but I've added to also disable if meter is not selected.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,192 @@
-				// TODO Using an alert is not the best. At some point this should be integrated
-				// with react.
-				showSuccessNotification('<h1>SUCCESS</h1>The meter upload was a success.');
+	// This is needed until a React CSV API is created or integrated with the Redux meter API
```
Would the team like to put in an issue about the needed CSV updates for React API? If not, I can do it.

mikedivine replied:
Sure, we will do this.

huss reviewed (changes requested):
Thanks for all the updates. I have made some comments to address. I'm also delaying final testing until this is done and development is merged. This is looking good and I think it should all be done soon.

huss committed:
fix pipeline testing & bug

- The pipeline tests were modified to use identifier and also for removal of the
createMeter option.
- uploadMeters still used name in some places even though identifier supplied so fixed.

huss commented:
I just pushed a commit to fix the pipeline test for identifier use and no create meter. It also fixes meter upload to consistently use identifier.

mikedivine committed:
updated server side upload Meters to send back the meter data that was added to the DB and update redux accordingly

mikedivine committed:
took away "access to uploadMeters" from CSV users

mikedivine committed:
updated some requested changes for CSV Upload of meters and readings

mikedivine committed:
using lodash for range function

mikedivine committed:
Merge remote-tracking branch 'upstream/development' into 1198-CSV

mikedivine committed:
Made sure CSV Readings state is using it's own readingsData state and having a chosen meter update the readingsData state.

Removed null options called 'meter' from certain Types such as TimeSortTypes and  BooleanMeterTypes which signaled to use the meter value or the default value. This is now the current way it is done even without these types.

Also updated some minor changes from review comment.s

mikedivine committed:
updated API return messages

mikedivine committed:
fixed csvPipeline Test Failures

huss commented on src/client/app/utils/api/UploadCSVApi.ts:
```diff
@@ -4,53 +4,97 @@
+		const { message, meters } = response;
+		const meterNames = meters.map(meter => meter.name).join(', ');
+		// If new meters were added to DB, have redux add them to getMeters state
+		/// If meters were updated to the DB, invalidate meters
```
extra /

mikedivine replied:
Removed.

huss commented on src/client/app/utils/api/UploadCSVApi.ts:
```diff
@@ -4,53 +4,97 @@
+};
+
+export const submitMeters = async (uploadPreferences: MetersCSVUploadPreferencesItem,
+	metersFile: File, dispatch: any): Promise<ApiResponse> => {
```
I'm curious if you think that dispatch should be passed or allocated here. In the calling code the dispatch is not used except to pass to this function. Same for submitReadings.

mikedivine replied:
This is not a component, that's why it is passed as a parameter

huss replied:
Needed since not a component.

huss commented on src/client/app/components/csv/MetersCSVUploadComponent.tsx:
```diff
@@ -4,107 +4,191 @@
-				showSuccessNotification('<h1>SUCCESS</h1>The meter upload was a success.');
+	const handleFileChange = (file: File | null) => {
+		setSelectedFile(file);
+		if (!file) {
```
I don't think there is anything wrong with this but I'm trying to figure out when it can happen. I tried Chrome and Firefox and I could not select a file that did not exist nor unset once set. Do you know if it is possible for the file not to exist? Applies to readings too.

mikedivine replied:
You are correct, think I was overdoing defensive programming.
I don't know if there any browsers that would let you put a filename in for a file that doesn't exist.
Removed Null and now doesn't check if File is Null.

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,513 @@
-	}
+	const handleSelectedMeterChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+		const selectedMeterIdentifier = e.target.value;
+		const foundMeter = visibleMeters.find(meter => meter.identifier === selectedMeterIdentifier) || null;
```
Since the meter came from a drop down menu, I'm trying to figure out how it would not be in visibleMeters?

mikedivine replied:
Was coding defensively, but I will get rid of the if statement and use Type Assertion. Is that what you meant?

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,513 @@
-	}
+	const handleNumberChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+		let aNumber = Number(e.target.value);
+		if (aNumber < 0) {
```
This may go away but I have a couple of thoughts:
1. The value is set to a different one that the user input without notice.
2. This uses the same name as other places but has modified functionality.
3. Given the min is 0 on the form, I cannot find a way to enter a negative value so is this needed?

mikedivine replied:
Before this I was able to enter a negative number by typing it in instead of using the arrows. I can put a warning when a user attempts to enter a negative number and I'll change the function name to handlePositiveNumberChange.  This could be done differently, so I'll check in with you before moving forward on this.

mikedivine replied:
We spoke, I will utilize what the other components use.

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,513 @@
+	// This is the state for the file to be uploaded
+	const [selectedFile, setSelectedFile] = React.useState<File | null>(null);
+	// This is the state for the chosen meter to upload readings for
+	const [selectedMeter, setSelectedMeter] = React.useState<MeterData>(defaultValues);
```
Can you help me understand why selectedMeter exists? From my quick check, it seems readingsData is also set with the updated meter info. Thus, is this a duplicated state?

mikedivine replied:
You are correct. This is not needed. I will remove this.

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,513 @@
+				setSelectedMeter(createdMeter);
+				setReadingsData(prevState => ({
+					...prevState,
+					meterIdentifier: createdMeter.identifier
```
A lot of other values are set below on meter change. Curious why not here?

mikedivine replied:
Yes, it should be, will add now.

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,513 @@
+						<FormGroup>
+							{isAdmin && <><CreateMeterModalComponent onCreateMeter={handleCreateMeter} /></>}
+						</FormGroup>
+						<FormGroup>
```
I'm really sorry to be bringing this up so late in the process. It is fine to either address or make an issue instead. I was looking at the meter create page and noticed it looks different than this page in terms of order, layout and checks. I was thinking the order should be similar to the meter create/edit page and the checks are the same (so you cannot enter the cumulative start/end if reset not chosen, etc.) I think making similar columns for the items would also be nice. I'm open to ideas on the exact layout (here there are section headers) but being consistent seems good.

mikedivine replied:
I will work on this.

mikedivine replied:
I matched the matching input fields in the same order as is on Create/Edit Meter.  I moved the file picker above the options and I moved the CSV file options directly under the file picker as it made the most sense to me.

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,513 @@
- * Returns a range of values between the specified lower and upper bounds.
- * @param lower The lower bound, which will be included in the range.
- * @param upper The upper bound, which will be excluded from the range.
- * @returns An array of values between starting from the lower bound and up to and excluding the upper bound.
+ * Defines the CSV Readings page
```
Blank line nice.

mikedivine replied:
Comment same as above, not a different line right?

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,513 @@
+import TooltipHelpComponent from '../TooltipHelpComponent';
+import TooltipMarkerComponent from '../TooltipMarkerComponent';
+import CreateMeterModalComponent from '../meters/CreateMeterModalComponent';
 /**
```
Blank line between is nice.

mikedivine replied:
Added.

huss commented on src/client/app/utils/api/UploadCSVApi.ts:
```diff
@@ -4,53 +4,97 @@
+		return { success: false, message: error.response.data };
+	}
+	dispatch(baseApi.util.invalidateTags(['Readings']));
+	return { success: true, message: message };
```
If I understand the logic, I think it is much clearer to put all the success in the try part. Then it catches all errors (not just the post) and it makes clear what goes with what. Using a return to avoid code should, in my opinion, only be used when necessary and where the logic would be difficult.

mikedivine replied:
I have moved it inside the try catch

huss commented on src/server/services/csvPipeline/uploadReadings.js:
```diff
@@ -126,43 +71,43 @@ async function uploadReadings(req, res, filepath, conn) {
 	}
 
-	if (timeSort === undefined || timeSort === TimeSortTypesJS.meter) {
+	if (timeSort === undefined) {
```
I think this works since the help page says "If a parameter is not supplied in a direct file upload then it defaults to the meter value.". However, I wanted to ask if you have tested this.

mikedivine replied:
Yes, I tested this using the curl command and didn't include parameters and it still worked.

huss replied:
Probably fine but can I get a confirmation on this?

mikedivine replied:
This has been tested, not sure if you will see this..

huss commented on src/server/test/web/csvPipelineTest.js:
```diff
@@ -518,23 +557,40 @@ for (let fileKey in testCases) {
-				expect(res).to.be.html;
 				// OED returns a string with messages that we check it is what was expected.
-				expect(res.text).to.equal(testCases[fileKey]['responseString'][index]);
+				try {
```
A few of thoughts on this:
1. The check for html was removed but a json check did not replace it.
2. The fact that the test needed to be modified is an indication that you changed the expected response. These responses have been published and sites have a reason to expect they will not change unless absolutely needed. They were set up as html for a reason and now this one is not html. I'm hoping the old behavior can easily be put back. Maybe you can not put a success code in the response as JSON and look at the response code. It looks as if that changes then the added item to via the JSON string is not needed.
3. Usually you have no need for a try/catch like this in a test. I think something unusual/undesirable is happening here.
I have not spent a lot of time on this so maybe I've missed something. We might need to discuss.

mikedivine replied:
Yes, reversing this back to previous per our discussion.

huss commented on src/server/routes/csv.js:
```diff
@@ -124,10 +123,13 @@ router.post('/meters', validateMetersCsvUploadParams, async (req, res) => {
-		success(req, res, 'Successfully inserted the meters.');
+		const newlyCreatedMeters = await uploadMeters(req, res, csvFilepath, conn);
+		// Send the newly created meters back to the frontend
+		res.status(200).send({ message: '<h1>SUCCESS</h1>Successfully inserted the meters.', meters: newlyCreatedMeters });
```
See a longer discussion in src/server/test/web/csvPipelineTest.js. I think this returned message changed too. It also changed how it is sent.

mikedivine replied:
Removed meterData being sent back as per discussion.

huss committed:
fix JSDoc

huss committed:
next version now 2.0.0

This is being done since the direct file upload params
are slightly changed and due to lots of other changes
outside this PR.

huss commented:
I'm noting that I pushed two commits.

huss reviewed (changes requested):
Thanks, again, to @mikedivine for continued work on this. Overall, this is looking good. I'm sorry to have more comments but please check them out. Let me know if I can help in any way.

mikedivine committed:
fix csv API - put code inside try catch

mikedivine committed:
updated changes to switch api back to not passing meters in. Fixed some state issues. Change Readings upload to closely mimick Create/Edit Meter layout.

mikedivine committed:
fixed import that wasn't used

mikedivine committed:
Merge branch 'OpenEnergyDashboard:development' into 1198-CSV

mikedivine committed:
fixed newMeter set issue

mikedivine committed:
updated to use invalidate on negative numbers

mikedivine committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into 1198-CSV

mikedivine committed:
accidentally removed a few lines, adding back in

mikedivine committed:
finished api return updates

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,521 @@
+								<FormGroup>
+									<Label for='cumulative'>
+										<div className='pb-1'>
+											{translate('csv.readings.param.cumulative')}
```
I noticed some of the titles differed from meter. I realized that there are a meter and csv equivalent keys in data.ts. I think it should be changed to only use a single set of keys. Ideally, the name would not be associated with one of these items (such as cumulative vs. meter.cumulative).

mikedivine replied:
I will go through these and match them and remove unused translations.

mikedivine replied:
completed

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,521 @@
+										value={readingsData.cumulative}
+										onChange={handleChange}
+									>
+										<option value={BooleanTypes.true}> {translate('BooleanMeterTypes.true')} </option>
```
Create meter looks over the enum to create the menu. While it is only two values, I think it is good practice. If you agree, could it mirror meter?

This applies to a couple of items.

mikedivine replied:
Yes, will mirror this style.

huss commented on src/client/app/components/csv/ReadingsCSVUploadComponent.tsx:
```diff
@@ -2,351 +2,521 @@
+										onChange={handleNumberChange}
+										invalid={readingsData.lengthGap < 0}
+									/>
+									<FormFeedback>
```
I'm okay with the feedback but I was unable to enter a negative value so I'm uncertain if the user can see these.

mikedivine replied:
Yes they don't allow a negative number, so we will remove the form feedback for these

huss commented on src/client/app/utils/api/UploadCSVApi.ts:
```diff
@@ -4,53 +4,75 @@
 } from '../../types/csvUploadForm';
+import ApiBackend from './ApiBackend';
+
+interface apiResponse {
```
I'm curious as to why this was changed from starting capitalized. A search found lots of interfaces in OED and I did not easily see one this way. What do you think?

mikedivine replied:
mistake, I agree it should be capitalized. Will change back.

huss commented on src/client/app/utils/api/UploadCSVApi.ts:
```diff
@@ -4,53 +4,75 @@
+		// Meter Data was sent to the DB, invalidate meters for now
+		dispatch(baseApi.util.invalidateTags(['MeterData']));
+		// meters were invalidated so all meter changes will now reflect in Redux state, now return
+		return { status: true, message: response };
```
Given this is used as a boolean where true is success, I'm curious about changing the name. Could you help me understand?

mikedivine replied:
No valid reason, think I was editing some stuff and must have thought it made more sense in the moment, but I agree 'success' makes more sense.  Will change it back.

huss reviewed (changes requested):
Thanks for another update. Testing did not find issues. I have a few more comments that I hope are easy to address. Other than that it seems ready.

mikedivine committed:
altered CSV users from having limited meterData to having full access to meterData to allow them to see meter defaults when uploading meter readings manually via CSV files

mikedivine committed:
Switched to use TrueFalseType which is used in Meter components instead of the exact same BooleanTypes mirroed in a different location.
Also fixed duplicated translations by removing them and altering CSV components to use the ones other components are using.

mikedivine committed:
fixed meters test since CSV users now also get all meterData along with Admin users

mikedivine committed:
Merge remote-tracking branch 'upstream/development' into 1198-CSV

mikedivine committed:
CSV Meters - If update is checked, then a Dropdown List of Meters is shown to choose which meter is updated

huss commented on src/client/app/utils/api/UploadCSVApi.ts:
```diff
@@ -4,53 +4,75 @@
+	// The Boolean values in state must be converted to the submitted values of yes and no.
+	const uploadPreferencesForm: ReadingsCSVUploadPreferencesForm = {
+		...uploadPreferences,
+		gzip: uploadPreferences.gzip ? TrueFalseType.true : TrueFalseType.false,
```
When I try to upload a file I get an enum error. I did not try to track down where it is coming from. Whatever changes are made, we also need to be sure that the curl commands work.

I also notice that BooleanTypes was removed from the client but is in many places in the server code. I supect this related.

mikedivine replied:
Yes, I removed BooleanTypes as Meter Components were using TrueFalseType which mirrored the BooleanTypes.  Curl method worked for me to add the meters while the Web Readings upload also worked but I did get the type error on the web CSV Meters upload attempt.  I'll try to figure what is happening.

mikedivine replied:
I am no longer able to reproduce this error. Let me know if you are still getting it.

huss replied:
This is the error I see:

![image](https://github.com/user-attachments/assets/09bfba56-83ec-492c-b79f-079e4a4ab0fa)

I still see the error and I think I have debugged it. You converted the client CSVUploads to be true/false. That maps true/false to yes/no which is good for the display. The issue is that the server still expected yes/no. See src/server/services/csvPipeline/validateCsvUploadParams.js for BooleanTypesJS which is used as the allowed values for cumulative, etc. I believe I did this a while back to show yes/no to the user but it was not the best way to do it. Now the enum loop shows that value but sends true/false. This is better since internally the value is a boolean. I verified this in two ways:

1. I put a console.log in src/server/services/csvPipeline/validateCsvUploadParams.js at the start of validateReadingsCsvUploadParams to show the values sent were true/false.
2. I changed `cumulative: new EnumParam('cumulative', [BooleanTypesJS.true, BooleanTypesJS.false]),` to `cumulative: Boolean,` and the error went away (there are other errors for the other params that have a similar issue.

Thus, I think this type of change will resolve the issue and improve the code from what it was before your changes. Please change for all the types needed, if you agree, and let me know if you think otherwise.

As long as I was deep in the code, I also wanted to double check about what happens if a direct (e.g., curl) request is made. In src/server/services/csvPipeline/uploadReadings.js on line 86 is checks if cumulative is not provided and then uses the meter value (assuming it exists). So, you can still get the meter value by not providing the value. This is what is already documented in https://openenergydashboard.org/helpV1_0_0/adminReadingsImport/ so no changes are needed. I'll add that what you said before about checking it was correct (I'm just paranoid :-).

huss replied:
I'm unresolving since I still see this. I was doing final testing when I saw this.

mikedivine replied:
Got it, I will work on fixing this.

mikedivine replied:
I have fixed this issue as far as I can tell. I created a utility function called normalizeBoolean which took any 'yes' 'no' values and converted to boolean.

huss reviewed (changes requested):
I was not sure if this was ready for review but I wanted to move it forward. I've made one comment to address and there are a couple of old ones to finalize. If you want to discuss the new one then let me know as it may not be trivial. Thanks for all your efforts.

mikedivine committed:
fixed returned csv error type issue, needed to return the error.message becayse the frontend CSV API is expecting a string back for the message.

mikedivine committed:
Revert "CSV Meters - fixed returned csv error type issue, needed to return the error.message becayse the frontend CSV API is expecting a string back for the message.

This reverts commit a9b8a2b66ea563ba7ef6c02e9102462b77e58f7b.

mikedivine reviewed (commented):
Oh I see I have to submit review for you to see comments

mikedivine reviewed (commented):
Oh I see I have to submit review for you to see comments

mikedivine committed:
Removed form feedback on csv readings that is unused.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/mikedivine/1284-6

huss commented:
I just merged in development since recent PRs caused a merge conflict and I had meant to do this PR earlier. I'll get it to work. After that there is one item unresolved that remains.

@mikedivine Can you look over the last comment and also verify my large merge did not cause an issue. My basic testing did not find any issues outside the last comment.

huss committed:
fix merge syntax mistake

huss committed:
fix meters test merge issues

- Incorrectly merged in some code.
- username vs email for user

huss commented:
I'm going to work with @mikedivine on the failed test to resolve meter name vs identifier.

huss committed:
meterName->meterIdentifier in pipeline tests

mikedivine committed:
Fixed CSV boolean paramter issues to allow for boolean, 'yes', 'no', 'true', 'false' so Curl command users can user either. Front End will now use strictly boolean values when sending to the CSV API. Also Fixed State issue on Readings CSV component.

mikedivine committed:
updated cumulative on Readings CSV component to disable it's associated fields when cumulative is set to 'no'.  This mirrors CreateMeter component

huss commented on src/server/services/csvPipeline/validateCsvUploadParams.js:
```diff
@@ -31,27 +19,30 @@ MeterTimeSortTypesJS = Object.freeze({
-});
+// This is to allow curl users to still send 'yes' and 'no' for parameters sent,
+// The frontend will just use standard boolean values to send to the API
+BooleanCheckArray = ['yes', 'no', 'true', 'false', true, false];
```
Should this be a const?

mikedivine replied:
yes, fixed

huss reviewed (commented):
In my attempt to keep up with the PRs, I reviewed your changes up to now. They all seem good. I made one small comment.

mikedivine committed:
finished changes to allow different boolean values such as 'yes' & 'no' & 'true' & 'false'

mikedivine committed:
put in backwards compatibility for accepting 'meterIdentifier' & 'meterName' using curl, tested and working

huss reviewed (approved):
I want to thank @mikedivine for a lot of work and effort to meet all the changing requests. I have reviewed the recent changes and retested. I found everything works as desired. This is ready to merge. Congratulations.
